File: System-Design/Geo-Spatial/Geohashing/geohash.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cassert>

using namespace std;

class Geohasher {
private:
    static constexpr const char* BASE32 = "0123456789bcdefghjkmnpqrstuvwxyz";

public:
    static string encode(double lat, double lon, int precision = 12) {
        vector<double> latInterval = {-90.0, 90.0};
        vector<double> lonInterval = {-180.0, 180.0};
        
        string geohash = "";
        vector<int> bits = {16, 8, 4, 2, 1};
        
        int bit = 0;
        int ch = 0;
        bool evenBit = true;

        while (geohash.length() < precision) {
            if (evenBit) {
                double mid = (lonInterval[0] + lonInterval[1]) / 2;
                if (lon > mid) {
                    ch |= bits[bit];
                    lonInterval[0] = mid;
                } else {
                    lonInterval[1] = mid;
                }
            } else {
                double mid = (latInterval[0] + latInterval[1]) / 2;
                if (lat > mid) {
                    ch |= bits[bit];
                    latInterval[0] = mid;
                } else {
                    latInterval[1] = mid;
                }
            }

            evenBit = !evenBit;

            if (bit < 4) {
                bit++;
            } else {
                geohash += BASE32[ch];
                bit = 0;
                ch = 0;
            }
        }

        return geohash;
    }
};

// --- CI/CD Automated Test ---
int main() {
    double lat = 37.8199;
    double lon = -122.4783;
    
    string result = Geohasher::encode(lat, lon, 9);
    
    assert(result == "9q8zh4yvc");
    
    cout << "C++ Geohashing Test Passed! (Result: " << result << ")\n";
    return 0;
}

File: System-Design/Geo-Spatial/Geohashing/geohash.py
class Geohasher:
    # Custom Base32 alphabet (omits 'a', 'i', 'l', 'o' to avoid visual confusion)
    BASE32 = "0123456789bcdefghjkmnpqrstuvwxyz"

    @staticmethod
    def encode(lat: float, lon: float, precision: int = 12) -> str:
        lat_interval = [-90.0, 90.0]
        lon_interval = [-180.0, 180.0]
        
        geohash = []
        bits = [16, 8, 4, 2, 1]
        
        bit = 0
        ch = 0
        even_bit = True # True for Longitude, False for Latitude

        while len(geohash) < precision:
            if even_bit:
                mid = (lon_interval[0] + lon_interval[1]) / 2
                if lon > mid:
                    ch |= bits[bit]
                    lon_interval[0] = mid
                else:
                    lon_interval[1] = mid
            else:
                mid = (lat_interval[0] + lat_interval[1]) / 2
                if lat > mid:
                    ch |= bits[bit]
                    lat_interval[0] = mid
                else:
                    lat_interval[1] = mid

            even_bit = not even_bit

            if bit < 4:
                bit += 1
            else:
                geohash.append(Geohasher.BASE32[ch])
                bit = 0
                ch = 0

        return "".join(geohash)

# --- CI/CD Automated Test ---
if __name__ == '__main__':
    # Coordinates for the Golden Gate Bridge, San Francisco
    lat = 37.8199
    lon = -122.4783
    
    # Standard 9-character precision (approx 4.7 x 4.7 meters)
    result = Geohasher.encode(lat, lon, precision=9)
    
    # Mathematical proof of correct encoding
    assert result == "9q8zh4yvc", f"Geohash failed! Expected '9q8zh4yvc', got '{result}'"
    
    print(f"Python Geohashing Test Passed! (Result: {result})")